ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImporterImple.java

```diff
@@ -167,11 +181,19 @@ public SubordinateTransaction getImportedTransaction(Xid xid)
 		if (xid == null)
 			throw new IllegalArgumentException();
 
-		SubordinateTransaction tx = _transactions.get(new SubordinateXidImple(xid));
+		AtomicReference<TransactionImple> holder = _transactions.get(new SubordinateXidImple(xid));
 
-		if (tx == null)
+		if (holder == null)
 			return null;
-		
+
+		SubordinateTransaction tx = holder.get();
+
+		if (tx == null) {
+			// another thread must be in the process of importing this transaction
+			// TODO the solution does not provide a mechanism to wait for the other thread to update the holder
```

Comment:
It shouldn't matter because the act of starting the transaction is atomic - that is, it is either observed to have happened, or observed to not have happened.  If the operation is in progress, then it hasn't happened yet.

However you can work around this by adding another synch block in here that re-gets the value and re-checks for null, if you want to guarantee that there is no transaction *and* there is no operation in progress to produce one.  This preserves the atomicity of the operation while also biasing this operation a little bit towards success.